In [22]:
from pathlib import Path

In [23]:
from pathlib import Path

BASE_DIR = Path.cwd()

if BASE_DIR.name == "notebooks":
    BASE_DIR = BASE_DIR.parent

RAW_DIR = BASE_DIR / "data" / "raw"
CLEAN_DIR = BASE_DIR / "data" / "clean"

print("Base folder:", BASE_DIR)
print("Raw folder:", RAW_DIR)
print("Clean folder:", CLEAN_DIR)
print("Raw folder exists:", RAW_DIR.exists())

Base folder: /Users/anastasiia/Downloads/bellabeat_github_portfolio_starter/bellabeat-case-study
Raw folder: /Users/anastasiia/Downloads/bellabeat_github_portfolio_starter/bellabeat-case-study/data/raw
Clean folder: /Users/anastasiia/Downloads/bellabeat_github_portfolio_starter/bellabeat-case-study/data/clean
Raw folder exists: True


In [24]:
print("Files in raw folder:")

for file in RAW_DIR.iterdir():
    print(file.name)

Files in raw folder:
dailyActivity_merged.csv
sleepDay_merged.csv


In [25]:
import pandas as pd

daily = pd.read_csv(RAW_DIR / "dailyActivity_merged.csv")
sleep = pd.read_csv(RAW_DIR / "sleepDay_merged.csv")

print("Files loaded successfully.")

Files loaded successfully.


In [26]:
daily.head()

,Id,ActivityDate,TotalSteps,TotalDistance,TrackerDistance,LoggedActivitiesDistance,VeryActiveDistance,ModeratelyActiveDistance,LightActiveDistance,SedentaryActiveDistance,VeryActiveMinutes,FairlyActiveMinutes,LightlyActiveMinutes,SedentaryMinutes,Calories
0,1503960366,3/25/2016,11004,7.11,7.11,0.0,2.57,0.46,4.07,0.0,33,12,205,804,1819
1,1503960366,3/26/2016,17609,11.55,11.55,0.0,6.92,0.73,3.91,0.0,89,17,274,588,2154
2,1503960366,3/27/2016,12736,8.53,8.53,0.0,4.66,0.16,3.71,0.0,56,5,268,605,1944
3,1503960366,3/28/2016,13231,8.93,8.93,0.0,3.19,0.79,4.95,0.0,39,20,224,1080,1932
4,1503960366,3/29/2016,12041,7.85,7.85,0.0,2.16,1.09,4.61,0.0,28,28,243,763,1886


In [27]:
sleep.head()

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
0,1503960366,4/12/2016 12:00:00 AM,1,327,346
1,1503960366,4/13/2016 12:00:00 AM,2,384,407
2,1503960366,4/15/2016 12:00:00 AM,1,412,442
3,1503960366,4/16/2016 12:00:00 AM,2,340,367
4,1503960366,4/17/2016 12:00:00 AM,1,700,712


In [28]:
from pathlib import Path

In [29]:
BASE_DIR = Path.cwd()

if BASE_DIR.name == "notebooks":
    BASE_DIR = BASE_DIR.parent

In [30]:
RAW_DIR = BASE_DIR / "data" / "raw"
CLEAN_DIR = BASE_DIR / "data" / "clean"

In [31]:
print("Base folder:", BASE_DIR)
print("Raw folder:", RAW_DIR)
print("Clean folder:", CLEAN_DIR)
print("Raw folder exists:", RAW_DIR.exists())

Base folder: /Users/anastasiia/Downloads/bellabeat_github_portfolio_starter/bellabeat-case-study
Raw folder: /Users/anastasiia/Downloads/bellabeat_github_portfolio_starter/bellabeat-case-study/data/raw
Clean folder: /Users/anastasiia/Downloads/bellabeat_github_portfolio_starter/bellabeat-case-study/data/clean
Raw folder exists: True


In [32]:
print("Daily rows and columns:" , daily.shape)
print("Sleep rows and columns:" , sleep.shape)

Daily rows and columns: (457, 15)
Sleep rows and columns: (413, 5)


## Initial Data Check

The daily activity dataset contains 457 rows and 15 columns.

The sleep dataset contains 413 rows and 5 columns.

Both datasets loaded correctly and are ready for the next processing checks.

In [33]:
daily.columns

Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='str')

In [34]:
sleep.columns

Index(['Id', 'SleepDay', 'TotalSleepRecords', 'TotalMinutesAsleep',
       'TotalTimeInBed'],
      dtype='str')

In [35]:
daily.info()

<class 'pandas.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        457 non-null    int64  
 1   ActivityDate              457 non-null    str    
 2   TotalSteps                457 non-null    int64  
 3   TotalDistance             457 non-null    float64
 4   TrackerDistance           457 non-null    float64
 5   LoggedActivitiesDistance  457 non-null    float64
 6   VeryActiveDistance        457 non-null    float64
 7   ModeratelyActiveDistance  457 non-null    float64
 8   LightActiveDistance       457 non-null    float64
 9   SedentaryActiveDistance   457 non-null    float64
 10  VeryActiveMinutes         457 non-null    int64  
 11  FairlyActiveMinutes       457 non-null    int64  
 12  LightlyActiveMinutes      457 non-null    int64  
 13  SedentaryMinutes          457 non-null    int64  
 14  Calories             

In [36]:
sleep.info()

<class 'pandas.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   Id                  413 non-null    int64
 1   SleepDay            413 non-null    str  
 2   TotalSleepRecords   413 non-null    int64
 3   TotalMinutesAsleep  413 non-null    int64
 4   TotalTimeInBed      413 non-null    int64
dtypes: int64(4), str(1)
memory usage: 16.3 KB


In [37]:
daily.duplicated().sum()

np.int64(0)

In [17]:
sleep.duplicated().sum()

np.int64(3)

In [39]:
sleep[sleep.duplicated(keep=False)]

,Id,SleepDay,TotalSleepRecords,TotalMinutesAsleep,TotalTimeInBed
160,4388161847,5/5/2016 12:00:00 AM,1,471,495
161,4388161847,5/5/2016 12:00:00 AM,1,471,495
222,4702921684,5/7/2016 12:00:00 AM,1,520,543
223,4702921684,5/7/2016 12:00:00 AM,1,520,543
379,8378563200,4/25/2016 12:00:00 AM,1,388,402
380,8378563200,4/25/2016 12:00:00 AM,1,388,402


The daily activity dataset does not contain duplicate rows.

The sleep dataset contains 3 duplicate rows. After reviewing them, I confirmed that they are exact duplicate records.

In [40]:
sleep_before = sleep.shape[0]

In [41]:
sleep= sleep.drop_duplicates()

In [42]:
sleep_after = sleep.shape[0]

In [43]:
print("Rows before:", sleep_before)

Rows before: 413


In [44]:
print("Rows after:", sleep_after)

Rows after: 410


In [46]:
print("Rows removed:", sleep_before -sleep_after)

Rows removed: 3


In [48]:
print("Sleep duplicates after cleaning:", sleep.duplicated().sum())

Sleep duplicates after cleaning: 0


I removed 3 duplicate rows from the sleep dataset.

The sleep dataset decreased from 413 rows to 410 rows.

After cleaning, the sleep dataset no longer contains duplicate records.

## Check Missing Values

In this step, I check if the datasets contain missing values.

Missing values can affect the analysis because some calculations may be incomplete or inaccurate.

In [49]:
daily.isnull().sum()

Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

In [50]:
sleep.isnull().sum()

Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

No missing values were found in the daily activity dataset or the sleep dataset.

In [54]:
daily_missing_rows = daily[daily.isnull().any(axis=1)]
sleep_missing_rows = sleep[sleep.isnull().any(axis=1)]

print("Daily rows with missing values:", daily_missing_rows.shape[0])
print("Sleep rows with missing values:", sleep_missing_rows.shape[0])

Daily rows with missing values: 0
Sleep rows with missing values: 0


I checked missing values by column and by row. This helped me confirm whether the datasets had incomplete records before analysis.

## Clean Column Names

In this step, I clean the column names to make them easier to use in Python.

I convert the column names to lowercase and separate words with underscores.

In [75]:
def clean_column_names(df):
    df = df.copy()
    df.columns = (
    df.columns
      .str.strip()
      .str.replace(" " , "_")
      .str.replace(r"(?<!^)(?=[A-Z])", "_", regex=True)
      .str.lower()           
                  
)
    return df

In [78]:
daily_test = clean_column_names(daily)
daily_test = clean_column_names(sleep)

In [86]:
daily = clean_column_names(daily)
sleep = clean_column_names(sleep)

In [ ]:
## Convert Date Columns

In this step, I convert the date columns from text format to datetime format.

This is important because dates are needed to analyze records over time and to connect activity and sleep data by user and date.

In [90]:
daily.columns

Index(['id', 'activity_date', 'total_steps', 'total_distance',
       'tracker_distance', 'logged_activities_distance',
       'very_active_distance', 'moderately_active_distance',
       'light_active_distance', 'sedentary_active_distance',
       'very_active_minutes', 'fairly_active_minutes',
       'lightly_active_minutes', 'sedentary_minutes', 'calories'],
      dtype='str')

In [91]:
sleep.columns

Index(['id', 'sleep_day', 'total_sleep_records', 'total_minutes_asleep',
       'total_time_in_bed'],
      dtype='str')

In [104]:
daily["activity_date"]= pd.to_datetime(
    daily["activity_date"],
    format = "%m,%d,%Y"

)
sleep["sleep_day"]= pd.to_datetime(
    sleep["sleep_day"],
    format= "%m,%d,%Y,%I,%M,%S,%p"

)

In [106]:
daily.info()

<class 'pandas.DataFrame'>
RangeIndex: 457 entries, 0 to 456
Data columns (total 15 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   id                          457 non-null    int64         
 1   activity_date               457 non-null    datetime64[us]
 2   total_steps                 457 non-null    int64         
 3   total_distance              457 non-null    float64       
 4   tracker_distance            457 non-null    float64       
 5   logged_activities_distance  457 non-null    float64       
 6   very_active_distance        457 non-null    float64       
 7   moderately_active_distance  457 non-null    float64       
 8   light_active_distance       457 non-null    float64       
 9   sedentary_active_distance   457 non-null    float64       
 10  very_active_minutes         457 non-null    int64         
 11  fairly_active_minutes       457 non-null    int64         
 12  light

In [107]:
sleep.info()

<class 'pandas.DataFrame'>
Index: 410 entries, 0 to 412
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   id                    410 non-null    int64         
 1   sleep_day             410 non-null    datetime64[us]
 2   total_sleep_records   410 non-null    int64         
 3   total_minutes_asleep  410 non-null    int64         
 4   total_time_in_bed     410 non-null    int64         
dtypes: datetime64[us](1), int64(4)
memory usage: 19.2 KB


In [109]:
daily.head()

,id,activity_date,total_steps,total_distance,tracker_distance,logged_activities_distance,very_active_distance,moderately_active_distance,light_active_distance,sedentary_active_distance,very_active_minutes,fairly_active_minutes,lightly_active_minutes,sedentary_minutes,calories
0,1503960366,2016-03-25,11004,7.11,7.11,0.0,2.57,0.46,4.07,0.0,33,12,205,804,1819
1,1503960366,2016-03-26,17609,11.55,11.55,0.0,6.92,0.73,3.91,0.0,89,17,274,588,2154
2,1503960366,2016-03-27,12736,8.53,8.53,0.0,4.66,0.16,3.71,0.0,56,5,268,605,1944
3,1503960366,2016-03-28,13231,8.93,8.93,0.0,3.19,0.79,4.95,0.0,39,20,224,1080,1932
4,1503960366,2016-03-29,12041,7.85,7.85,0.0,2.16,1.09,4.61,0.0,28,28,243,763,1886


In [110]:
sleep.head()

,id,sleep_day,total_sleep_records,total_minutes_asleep,total_time_in_bed
0,1503960366,2016-04-12,1,327,346
1,1503960366,2016-04-13,2,384,407
2,1503960366,2016-04-15,1,412,442
3,1503960366,2016-04-16,2,340,367
4,1503960366,2016-04-17,1,700,712


## Create Date Column

I create a new date column in both datasets.

This will help me join activity and sleep records by user and date later in the analysis.

In [111]:
daily["date"] = daily["activity_date"].dt.date
sleep["date"] = sleep["sleep_day"].dt.date

In [114]:
daily[["id", "activity_date", "date"]].head()

,id,activity_date,date
0,1503960366,2016-03-25,2016-03-25
1,1503960366,2016-03-26,2016-03-26
2,1503960366,2016-03-27,2016-03-27
3,1503960366,2016-03-28,2016-03-28
4,1503960366,2016-03-29,2016-03-29


In [115]:
sleep[["id", "sleep_day", "date"]].head()

,id,sleep_day,date
0,1503960366,2016-04-12,2016-04-12
1,1503960366,2016-04-13,2016-04-13
2,1503960366,2016-04-15,2016-04-15
3,1503960366,2016-04-16,2016-04-16
4,1503960366,2016-04-17,2016-04-17


The date column was created in both datasets.

This gives both tables the same date field, which will make it easier to join activity and sleep data by user and date during the analysis phase.

## Create Total Active Minutes

I create a new column called total_active_minutes.

This column combines very active, fairly active, and lightly active minutes into one total activity measure.

In [116]:
daily["total_active_minutes"] = (
    daily["very_active_minutes"]
    + daily["fairly_active_minutes"]
    + daily["lightly_active_minutes"]
)

In [117]:
daily[[
    "id",
    "date",
    "very_active_minutes",
    "fairly_active_minutes",
    "lightly_active_minutes",
    "total_active_minutes"
]].head()

,id,date,very_active_minutes,fairly_active_minutes,lightly_active_minutes,total_active_minutes
0,1503960366,2016-03-25,33,12,205,250
1,1503960366,2016-03-26,89,17,274,380
2,1503960366,2016-03-27,56,5,268,329
3,1503960366,2016-03-28,39,20,224,283
4,1503960366,2016-03-29,28,28,243,299


## Create Activity Level

I create an activity_level column based on total daily steps.

This will help classify each day as Low, Medium, or High activity.

In [118]:
def activity_level(steps):
    if steps < 5000:
        return "Low"
    elif steps < 10000:
        return "Medium"
    else:
        return "High"

In [119]:
daily["activity_level"] = daily["total_steps"].apply(activity_level)

In [120]:
daily[[
    "id",
    "date",
    "total_steps",
    "activity_level"
]].head(10)

,id,date,total_steps,activity_level
0,1503960366,2016-03-25,11004,High
1,1503960366,2016-03-26,17609,High
2,1503960366,2016-03-27,12736,High
3,1503960366,2016-03-28,13231,High
4,1503960366,2016-03-29,12041,High
5,1503960366,2016-03-30,10970,High
6,1503960366,2016-03-31,12256,High
7,1503960366,2016-04-01,12262,High
8,1503960366,2016-04-02,11248,High
9,1503960366,2016-04-03,10016,High


In [121]:
daily["activity_level"].value_counts()

activity_level
Low       197
Medium    133
High      127
Name: count, dtype: int64

I created total_active_minutes to combine all active minutes into one column.

I also created activity_level to classify daily step counts as Low, Medium, or High activity.

These new columns will help analyze activity patterns more clearly.

In [122]:
daily.head()

,id,activity_date,total_steps,total_distance,tracker_distance,logged_activities_distance,very_active_distance,moderately_active_distance,light_active_distance,sedentary_active_distance,very_active_minutes,fairly_active_minutes,lightly_active_minutes,sedentary_minutes,calories,date,total_active_minutes,activity_level
0,1503960366,2016-03-25,11004,7.11,7.11,0.0,2.57,0.46,4.07,0.0,33,12,205,804,1819,2016-03-25,250,High
1,1503960366,2016-03-26,17609,11.55,11.55,0.0,6.92,0.73,3.91,0.0,89,17,274,588,2154,2016-03-26,380,High
2,1503960366,2016-03-27,12736,8.53,8.53,0.0,4.66,0.16,3.71,0.0,56,5,268,605,1944,2016-03-27,329,High
3,1503960366,2016-03-28,13231,8.93,8.93,0.0,3.19,0.79,4.95,0.0,39,20,224,1080,1932,2016-03-28,283,High
4,1503960366,2016-03-29,12041,7.85,7.85,0.0,2.16,1.09,4.61,0.0,28,28,243,763,1886,2016-03-29,299,High


## Create Sleep Columns

I create new sleep columns to make the sleep data easier to analyze.

The original sleep values are in minutes, so I convert them to hours and calculate the time spent in bed while awake.

In [124]:
sleep["sleep_hours"] = sleep["total_minutes_asleep"] / 60

In [125]:
sleep["time_in_bed_hours"] = sleep["total_time_in_bed"] / 60

In [126]:
sleep["awake_time_in_bed_minutes"] = (
    sleep["total_time_in_bed"] - sleep["total_minutes_asleep"]
)

In [127]:
sleep[[
    "id",
    "date",
    "total_minutes_asleep",
    "total_time_in_bed",
    "sleep_hours",
    "time_in_bed_hours",
    "awake_time_in_bed_minutes"
]].head()

,id,date,total_minutes_asleep,total_time_in_bed,sleep_hours,time_in_bed_hours,awake_time_in_bed_minutes
0,1503960366,2016-04-12,327,346,5.450000,5.766667,19
1,1503960366,2016-04-13,384,407,6.400000,6.783333,23
2,1503960366,2016-04-15,412,442,6.866667,7.366667,30
3,1503960366,2016-04-16,340,367,5.666667,6.116667,27
4,1503960366,2016-04-17,700,712,11.666667,11.866667,12


In [129]:
sleep["sleep_hours"] = (sleep["total_minutes_asleep"] / 60).round(2)
sleep["time_in_bed_hours"] = (sleep["total_time_in_bed"] / 60).round(2)

In [130]:
sleep.head()

,id,sleep_day,total_sleep_records,total_minutes_asleep,total_time_in_bed,date,sleep_hours,time_in_bed_hours,awake_time_in_bed_minutes
0,1503960366,2016-04-12,1,327,346,2016-04-12,5.45,5.77,19
1,1503960366,2016-04-13,2,384,407,2016-04-13,6.40,6.78,23
2,1503960366,2016-04-15,1,412,442,2016-04-15,6.87,7.37,30
3,1503960366,2016-04-16,2,340,367,2016-04-16,5.67,6.12,27
4,1503960366,2016-04-17,1,700,712,2016-04-17,11.67,11.87,12


I converted sleep minutes and time in bed minutes into hours to make the values easier to understand. I also calculated awake_time_in_bed_minutes as the difference between time in bed and time asleep.

In [131]:
print("Daily shape:", daily.shape)
print("Sleep shape:", sleep.shape)

Daily shape: (457, 18)
Sleep shape: (410, 9)


In [132]:
print("Daily duplicates:", daily.duplicated().sum())
print("Sleep duplicates:", sleep.duplicated().sum())

Daily duplicates: 0
Sleep duplicates: 0


In [133]:
print("Daily missing values:")
print(daily.isnull().sum())

print("\nSleep missing values:")
print(sleep.isnull().sum())

Daily missing values:
id                            0
activity_date                 0
total_steps                   0
total_distance                0
tracker_distance              0
logged_activities_distance    0
very_active_distance          0
moderately_active_distance    0
light_active_distance         0
sedentary_active_distance     0
very_active_minutes           0
fairly_active_minutes         0
lightly_active_minutes        0
sedentary_minutes             0
calories                      0
date                          0
total_active_minutes          0
activity_level                0
dtype: int64

Sleep missing values:
id                           0
sleep_day                    0
total_sleep_records          0
total_minutes_asleep         0
total_time_in_bed            0
date                         0
sleep_hours                  0
time_in_bed_hours            0
awake_time_in_bed_minutes    0
dtype: int64


In [134]:
daily.columns

Index(['id', 'activity_date', 'total_steps', 'total_distance',
       'tracker_distance', 'logged_activities_distance',
       'very_active_distance', 'moderately_active_distance',
       'light_active_distance', 'sedentary_active_distance',
       'very_active_minutes', 'fairly_active_minutes',
       'lightly_active_minutes', 'sedentary_minutes', 'calories', 'date',
       'total_active_minutes', 'activity_level'],
      dtype='str')

In [135]:
sleep.columns

Index(['id', 'sleep_day', 'total_sleep_records', 'total_minutes_asleep',
       'total_time_in_bed', 'date', 'sleep_hours', 'time_in_bed_hours',
       'awake_time_in_bed_minutes'],
      dtype='str')

In [136]:
CLEAN_DIR.mkdir(parents=True, exist_ok=True)

In [137]:
daily.to_csv(CLEAN_DIR / "daily_activity_clean.csv", index=False)
sleep.to_csv(CLEAN_DIR / "sleep_clean.csv", index=False)

print("Clean files saved successfully.")

Clean files saved successfully.


In [138]:
for file in CLEAN_DIR.iterdir():
    print(file.name)

sleep_clean.csv
daily_activity_clean.csv


## Process Summary

In this phase, I cleaned and prepared the Fitbit activity and sleep datasets for analysis.

The main steps were:

- Loaded the raw CSV files
- Checked rows and columns
- Reviewed column names and data types
- Checked duplicate records
- Removed 3 duplicate rows from the sleep dataset
- Checked missing values by column and by row
- Cleaned column names
- Converted date columns to datetime format
- Created a common date column
- Created new calculated columns
- Exported clean CSV files

The cleaned datasets are ready for the analysis phase.